### PR TITLE
fix(mfa): Add types to `allauth.mfa` flows

### PR DESCRIPTION
**Pull request by akx**
This PR adds a bunch of types to `allauth.mfa`'s internals.

Our app needs to set up MFA with a slightly custom flow, so we can't use the forms or headless features, so I needed to delve a bit into how things work under the hood, and figured it'd be useful if types were documented technically too, for IDEs etc.

Feels like `TOTP` and `RecoveryCodes` should have a superclass, but that's beyond the scope of this PR...

### CODE REVIEW

File: allauth/mfa/models.py [hunk 1]
```diff
@@ -14,7 +14,7 @@
 
 
 class AuthenticatorManager(models.Manager):
-    def delete_dangling_recovery_codes(self, user):
+    def delete_dangling_recovery_codes(self, user) -> "Authenticator":
```
Comment (pennersr) [inline allauth/mfa/models.py:17]: **Issue**: This can also return `None`.

File: allauth/mfa/recovery_codes.py [hunk 2]
```diff
@@ -79,7 +80,7 @@ def get_unused_codes(self):
             ret.append(code)
         return ret
 
-    def _validate_migrated_code(self, code):
+    def _validate_migrated_code(self, code: str) -> bool:
```
Comment (pennersr) [inline allauth/mfa/recovery_codes.py:83]: **Issue**: Returns `None` as well.

**Review by pennersr — CHANGES_REQUESTED** [review]

**Comment by akx** [comment]
> Can you please rebase? Thanks.

Done, let's see if it passes! 🤞 

> adding type hints without enforcing them easily leads to errors.

Tangentially: I know we've talked about Ruff before and I know how you feel about it, but it was pretty useful here, as the [ANN series](https://docs.astral.sh/ruff/rules/#flake8-annotations-ann) of rules can narrow down places that are missing annotations, and a human can then figure them out.

**Comment by pennersr** [comment]
See #3905 -- some of your changes are not Python 3.7 compatible, that will be addressed by #3905.

**Comment by akx** [comment]
@pennersr On that note, what's your stance on using `from __future__ import annotations` and futuristic annotations (e.g. `list[Foo] | None`)? I don't think anything in allauth is expecting to introspect annotations (where unevaluated futured annotations could be an issue).

**Pull request merged**